### PR TITLE
Make @objcImpl work with @_cdecl

### PR DESCRIPTION
**Pull request by beccadax**
Swift can now emit C functions which match a global function declaration in a C header. That is, you can apply  `@_objcImplementation` to a `@_cdecl` function and it will match it to a global C function in a header.

This doesn't yet have all the functionality I'd like—I'm hoping to cover computed properties and at least some import-as-member functions too. But it already has enough functionality to be useful.

Fixes rdar://103845088.

### CODE REVIEW

**Comment by beccadax** [comment]
@swift-ci please test

**Comment by beccadax** [comment]
@swift-ci please test

**Comment by beccadax** [comment]
@swift-ci please test windows platform

**Comment by beccadax** [comment]
Note that I got a pass on Windows yesterday with the exact same code: https://ci-external.swift.org/job/swift-PR-windows/19506/

**Comment by beccadax** [comment]
@swift-ci please test windows platform

**Comment by beccadax** [comment]
@swift-ci please test

File: test/Interpreter/objc_implementation.swift [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// RUN: %target-run-simple-swift(-import-objc-header %S/Inputs/objc_implementation.h -D TOP_LEVEL_CODE -swift-version 5) %s | %FileCheck %s
+// RUN: %target-run-simple-swift(-import-objc-header %S/Inputs/objc_implementation.h -D TOP_LEVEL_CODE -swift-version 5) %s -save-temps | %FileCheck %s
```
Comment (tshortli) [inline test/Interpreter/objc_implementation.swift:1]: Is leaving `-save-temps` intentional here?

**Review by tshortli — APPROVED** [review]

File: test/Interpreter/objc_implementation.swift (hunk 1 above)
Comment (beccadax) [reply test/Interpreter/objc_implementation.swift]: Nope—I've now removed it. Thanks!

**Comment by beccadax** [comment]
@swift-ci please test and merge

**Comment by beccadax** [comment]
@swift-ci please test and merge

**Comment by beccadax** [comment]
@swift-ci please smoke test and merge

**Comment by beccadax** [comment]
@swift-ci please smoke test

File: lib/AST/Decl.cpp [hunk 2]
```diff
@@ -3869,6 +3869,22 @@ void ValueDecl::setInterfaceType(Type type) {
...
+          && clangDecl->getIdentifier())
+      return clangDecl->getName();
+  }
+
+  // Handle explicit cdecl attributes.
+  if (auto cdeclAttr = getAttrs().getAttribute<CDeclAttr>()) {
+    return cdeclAttr->Name;
+  }
+
+  return "";
```
Comment (DougGregor) [inline lib/AST/Decl.cpp:3885]: There's also `@extern(c, "name")`, IIRC

Comment (beccadax) [reply lib/AST/Decl.cpp:3885]: It looks like `@_extern` requires the function to not have a body, so it wouldn't make sense to use it with `@implementation` (which provides a body). Is that changing?

**Pull request merged**